bin/validate.py

```diff
@@ -180,3 +180,79 @@ def run(self, testcase, run=None, constraints=None, args=None):
                 cwd=run.feedbackdir,
                 timeout=config.get_timeout(),
             )
+
+# Testcase input is only allowed to contain newlines and printable characters
+def _is_invalid_input_byte(byte):
+    if byte == ord('\n'):
+        return False
+    if byte >= 0x20 and byte < 0x7F:
+        return False
+    return True
+
+
+# User output is additionally allowed to contain all other types of whitespaces
+def _is_invalid_output_byte(byte):
+    if byte == ord('\t'):
+        return False
+    if byte == ord('\r'):
+        return False
+    if byte == ord('\v'):
+        return False
+    if byte == ord('\f'):
+        return False
+    return _is_invalid_input_byte(byte)
+
+
+# assumes that the only possible whitespaces are space and newline
+# allows \n\n
+def _has_consecutive_whitespaces(bytes):
+    last = -1
+    for byte in bytes:
+        cur_whitespace = byte == ord(' ') or byte == ord('\n')
+        if last == ord(' ') and cur_whitespace:
+            return True
+        if last == ord('\n') and byte == ord(' '):
+            return True
+        last = byte
+    return False
+
+def generic_validation(validator_type, file, *, bar, sample_ans=False):
```

Comment:
1. Add some comment on the purpose of this function. 'generic' wasn't super obvious to me on first reading.
2. Comment that `--skip-testcase-sanity-checks` disables this.
3. Instead of using `input` and `output` for `validator_type`, and `sample_ans`, it may be more consistent to use `input_format`, `output_format`, and `output`.